ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt

```diff
@@ -130,37 +182,52 @@ public class HttpRequestRetry internal constructor(configuration: Configuration)
          * Specifies delay logic for retries. The [block] accepts the number of retries
          * and should return the number of milliseconds to wait before retrying.
          */
-        public fun delayMillis(block: (retry: Int) -> Long) {
-            delayMillis = block
+        public fun delayMillis(
+            respectRetryAfterHeader: Boolean = true,
+            block: DelayContext.(retry: Int) -> Long
+        ) {
+            delayMillis = {
+                val retryAfter = response?.headers?.get(HttpHeaders.RetryAfter)?.toLong()
+                if (respectRetryAfterHeader && retryAfter != null) {
```

Comment:
This should be `maxOf(block(), retryAfter ?: 0)` or similar.  The server might not know you are in exponential fall-off and it is try to pause clients or slow them down, but at the same time your logic of falling off should allow to be even slower since the client has more context about repeated calls than the server.